<a href="https://colab.research.google.com/github/ddoyungyung/Final-Team5/blob/main/%EB%A9%8B%EC%9F%81%EC%9D%B4_%EC%82%AC%EC%9E%90%EC%B2%98%EB%9F%BC_final_%ED%94%84%EB%A1%9C%EC%A0%9D%ED%8A%B8_%EC%86%8C%EC%9C%A41.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

마운트 및 라이브러리 불러오기

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
import numpy as np

csv 불러오기

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Hotel_Reviews.csv')

In [ ]:
df

Hotel_Address  \
0        s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
1        s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
2        s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
3        s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
4        s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
...                                                   ...   
515733  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...   
515734  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...   
515735  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...   
515736  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...   
515737  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...   

        Additional_Number_of_Scoring Review_Date  Average_Score  \
0                                194    8/3/2017            7.7   
1                                194    8/3/2017            7.7   
2                                194   7/31/2017            7.7   
3                                194   7/31/2017            7.7   
4                                194   7/24/2017            7.7   
...                              ...         ...            ...   
515733                           168   8/30/2015            8.1   
515734                           168   8/22/2015            8.1   
515735                           168   8/19/2015            8.1   
515736                           168   8/17/2015            8.1   
515737                           168    8/9/2015            8.1   

                   Hotel_Name Reviewer_Nationality  \
0                 Hotel Arena              Russia    
1                 Hotel Arena             Ireland    
2                 Hotel Arena           Australia    
3                 Hotel Arena      United Kingdom    
4                 Hotel Arena         New Zealand    
...                       ...                  ...   
515733  Atlantis Hotel Vienna              Kuwait    
515734  Atlantis Hotel Vienna             Estonia    
515735  Atlantis Hotel Vienna               Egypt    
515736  Atlantis Hotel Vienna              Mexico    
515737  Atlantis Hotel Vienna             Hungary    

                                          Negative_Review  \
0        I am so angry that i made this post available...   
1                                             No Negative   
2        Rooms are nice but for elderly a bit difficul...   
3        My room was dirty and I was afraid to walk ba...   
4        You When I booked with your company on line y...   
...                                                   ...   
515733   no trolly or staff to help you take the lugga...   
515734           The hotel looks like 3 but surely not 4    
515735   The ac was useless It was a hot week in vienn...   
515736                                        No Negative   
515737       I was in 3rd floor It didn t work Free Wife    

        Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
0                                     397                     1403   
1                                       0                     1403   
2                                      42                     1403   
3                                     210                     1403   
4                                     140                     1403   
...                                   ...                      ...   
515733                                 14                     2823   
515734                                 11                     2823   
515735                                 19                     2823   
515736                                  0                     2823   
515737                                 13                     2823   

                                          Positive_Review  \
0        Only the park outside of the hotel was beauti...   
1        No real complaints the hotel was great great ...   
2        Location was good and staff were ok It is cut...   
3        Great location in nice surroun

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 515738 entries, 0 to 515737
Data columns (total 17 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   Hotel_Address                               515738 non-null  object 
 1   Additional_Number_of_Scoring                515738 non-null  int64  
 2   Review_Date                                 515738 non-null  object 
 3   Average_Score                               515738 non-null  float64
 4   Hotel_Name                                  515738 non-null  object 
 5   Reviewer_Nationality                        515738 non-null  object 
 6   Negative_Review                             515738 non-null  object 
 7   Review_Total_Negative_Word_Counts           515738 non-null  int64  
 8   Total_Number_of_Reviews                     515738 non-null  int64  
 9   Positive_Review                             515738 non-null  object 
 

In [ ]:
df.head()

Hotel_Address  \
0   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
1   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
2   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
3   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
4   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   

   Additional_Number_of_Scoring Review_Date  Average_Score   Hotel_Name  \
0                           194    8/3/2017            7.7  Hotel Arena   
1                           194    8/3/2017            7.7  Hotel Arena   
2                           194   7/31/2017            7.7  Hotel Arena   
3                           194   7/31/2017            7.7  Hotel Arena   
4                           194   7/24/2017            7.7  Hotel Arena   

  Reviewer_Nationality                                    Negative_Review  \
0              Russia    I am so angry that i made this post available...   
1             Ireland                                         No Negative   
2           Australia    Rooms are nice but for elderly a bit difficul...   
3      United Kingdom    My room was dirty and I was afraid to walk ba...   
4         New Zealand    You When I booked with your company on line y...   

   Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
0                                397                     1403   
1                                  0                     1403   
2                                 42                     1403   
3                                210                     1403   
4                                140                     1403   

                                     Positive_Review  \
0   Only the park outside of the hotel was beauti...   
1   No real complaints the hotel was great great ...   
2   Location was good and staff were ok It is cut...   
3   Great location in nice surroundings the bar a...   
4    Amazing location and building Romantic setting    

   Review_Total_Positive_Word_Counts  \
0                                 11   
1                                105   
2                                 21   
3                                 26   
4                                  8   

   Total_Number_of_Reviews_Reviewer_Has_Given  Reviewer_Score  \
0                                           7             2.9   
1                                           7             7.5   
2                                           9             7.1   
3                                           1             3.8   
4                                           3             6.7   

                                                Tags days_since_review  \
0  [' Leisure trip ', ' Couple ', ' Duplex Double...            0 days   
1  [' Leisure trip ', ' Couple ', ' Duplex Double...            0 days   
2  [' Leisure trip ', ' Family with young childre...            3 days   
3  [' Leisure trip ', ' Solo traveler ', ' Duplex...            3 days   
4  [' Leisure trip ', ' Couple ', ' Suite ', ' St...           10 days   

         lat       lng  
0  52.360576  4.915968  
1  52.360576  4.915968  
2  52.360576  4.915968  
3  52.360576  4.915968  
4  52.360576  4.915968

```
데이터 개수: 515,783
0-14행까지는 결측치 존재x
15,16번 행은 각각 3268개의 결측치 존재
```

------------------------------------------

Reviwer_counts 컬럼의 값별 개수 확인하기

In [ ]:
score_counts = df['Reviewer_Score'].value_counts().sort_index()
score_counts

Reviewer_Score
2.5       2199
2.9       1589
3.0         37
3.1          8
3.3       2772
3.5         79
3.8       4044
4.0         84
4.2       5095
4.4         12
4.5        122
4.6       6240
5.0       8422
5.4       9526
5.5        227
5.6         17
5.8      12034
6.0        253
6.3      14924
6.5        345
6.7      18793
6.9         29
7.0        411
7.1      24855
7.5      34878
7.9      33097
8.0        505
8.1         39
8.3      41128
8.5        508
8.8      46429
9.0        647
9.2      58583
9.4         63
9.5        724
9.6      71167
10.0    115853
Name: count, dtype: int64

```
10점이 115953으로 가장 많이 존재

하지만 특정 숫자를 기준으로 두고 하기엔 쉽지 않음.(편향이 되어 있으므로)
```

기준이 될 평점 고르기

In [ ]:
df['Score_Difference'] = df['Reviewer_Score'] - df['Average_Score']

In [ ]:
df[['Reviewer_Score', 'Average_Score', 'Score_Difference']]

Reviewer_Score  Average_Score  Score_Difference
0                  2.9            7.7              -4.8
1                  7.5            7.7              -0.2
2                  7.1            7.7              -0.6
3                  3.8            7.7              -3.9
4                  6.7            7.7              -1.0
...                ...            ...               ...
515733             7.0            8.1              -1.1
515734             5.8            8.1              -2.3
515735             2.5            8.1              -5.6
515736             8.8            8.1               0.7
515737             8.3            8.1               0.2

[515738 rows x 3 columns]

In [ ]:
df['Score_Difference'].describe()

count    515738.00000
mean         -0.00241
std           1.52605
min          -7.00000
25%          -0.80000
50%           0.40000
75%           1.10000
max           3.60000
Name: Score_Difference, dtype: float64

- 데이터는 하나의 호텔에 대한 것이 아니라, 유럽전역에 있는 여러 호텔에 대한 것임.
- 그 호텔들의 평균 점수가 7.7이므로 Score_Difference값이 2.3을 초과하는 값(10-7.7)이 나오는 게 당연함!
- Score_Difference <0 , Reviewer_Score < Average_Score
- Score_Difference >0 , Reviewer_Score > Average_Score

기준: Score_Difference <0

In [ ]:
low_score_df = df[df['Score_Difference'] < 0]

In [ ]:
low_score_df.head()

Hotel_Address  \
0   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
1   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
2   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
3   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
4   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   

   Additional_Number_of_Scoring Review_Date  Average_Score   Hotel_Name  \
0                           194    8/3/2017            7.7  Hotel Arena   
1                           194    8/3/2017            7.7  Hotel Arena   
2                           194   7/31/2017            7.7  Hotel Arena   
3                           194   7/31/2017            7.7  Hotel Arena   
4                           194   7/24/2017            7.7  Hotel Arena   

  Reviewer_Nationality                                    Negative_Review  \
0              Russia    I am so angry that i made this post available...   
1             Ireland                                         No Negative   
2           Australia    Rooms are nice but for elderly a bit difficul...   
3      United Kingdom    My room was dirty and I was afraid to walk ba...   
4         New Zealand    You When I booked with your company on line y...   

   Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
0                                397                     1403   
1                                  0                     1403   
2                                 42                     1403   
3                                210                     1403   
4                                140                     1403   

                                     Positive_Review  \
0   Only the park outside of the hotel was beauti...   
1   No real complaints the hotel was great great ...   
2   Location was good and staff were ok It is cut...   
3   Great location in nice surroundings the bar a...   
4    Amazing location and building Romantic setting    

   Review_Total_Positive_Word_Counts  \
0                                 11   
1                                105   
2                                 21   
3                                 26   
4                                  8   

   Total_Number_of_Reviews_Reviewer_Has_Given  Reviewer_Score  \
0                                           7             2.9   
1                                           7             7.5   
2                                           9             7.1   
3                                           1             3.8   
4                                           3             6.7   

                                                Tags days_since_review  \
0  [' Leisure trip ', ' Couple ', ' Duplex Double...            0 days   
1  [' Leisure trip ', ' Couple ', ' Duplex Double...            0 days   
2  [' Leisure trip ', ' Family with young childre...            3 days   
3  [' Leisure trip ', ' Solo traveler ', ' Duplex...            3 days   
4  [' Leisure trip ', ' Couple ', ' Suite ', ' St...           10 days   

         lat       lng  Score_Difference  
0  52.360576  4.915968              -4.8  
1  52.360576  4.915968              -0.2  
2  52.360576  4.915968              -0.6  
3  52.360576  4.915968              -3.9  
4  52.360576  4.915968              -1.0

In [ ]:
print(len(df))
print(len(low_score_df))

515738
201512


음수리뷰: 전체 총 515,738 리뷰 중 201,512(약 40%)

양수리뷰: 전체 총 515,783 리뷰 중 314,226(약 60%)

-> 음수리뷰: 왜 불만족했을까 분석

-> 양수리뷰: 무엇때문에 만족을 했는가 분석


------------------------------------

## 전처리

1. 텍스트 전처리

```
특이한 변수: Negative_Review
"If the reviewer did not write a negative review, it is represented as "No Negative"

따라서, 리뷰어가 긍정적인 리뷰만 남기고 부정적인 리뷰칸을 비워두었다(부정적인 리뷰를 작성하지 않았다) == No Negative로 표시!
```

In [ ]:
import re
import string

def clean_text(text):
    if not isinstance(text, str): # 문자열이 아닐 경우 빈 문자열 반환
        return ''

    if 'no negative' in text.lower():
        return ''
    text = re.sub(r'\[.*?\]', '', text) #re.sub : 정규표현식을 이용해 특정 패턴의 문자열을 다른 문자열로 대체
    text = re.sub(r'https?://\S+|www\.\S+', '', text)
    text = re.sub(r'[%s]' % re.escape(string.punctuation), '', text)
    #string.punctuation : string라이브러리에 미리 정의된 구두점들
    #re.escape():파이썬 자체 내장 구두점들을 특수문자가 아닌 일반문자로 인식하도록
    text = re.sub(r'\w*\d\w*', '', text)
    #\w*\d\w*: 숫자 앞뒤로 단어가 붙어있는 경우 제거 예:room1,2nd,,,
    text = re.sub(r'\n', '', text)
    text = re.sub(r'\s+', ' ', text).strip()
    text = text.lower()

    return text

In [ ]:
df['Cleaned_Negative_Review'] = df['Negative_Review'].apply(clean_text)
df['Score_Category'] = df['Score_Difference'].apply(lambda x: 'Low_Score' if x < 0 else 'High_Score')

In [ ]:
low_score_df = df[df['Score_Difference'] < 0].copy()
low_score_df[['Negative_Review', 'Cleaned_Negative_Review']].head().to_string()

'                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       

---------------------------------

가설1:"단순히 단어의 사용 빈도보다,'소음이 시끄럽다'와 같이 특정 대상(명사)과 감정(형용사)이 결합된 표현이 낮은 평점에 더 큰 영향을 미칠 것이다."

NLP(자연어처리) 라이브러리 설치
* 단순한 단어의 나열이 아닌, 컴퓨터가 그 단어의 품사를 인식하게 하기 위해 사용

명사-형용사 쌍 추출 및 빈도 분석

In [ ]:
import spacy
from collections import Counter #Counter: 리스트 안에 있는 요소들의 빈도수를 계산

1차 추출

In [ ]:
#보류코드, 사용 X

nlp = spacy.load("en_core_web_sm") #spacy.load:영어문장을 분석할 수 있는 **전문가**를 데려오는 코드

def extract_noun_adj_pairs(text):
    pairs = []
    if not text or text.isspace():
        return pairs

    doc = nlp(text) #nlp가 text를 분석

    for token in doc: #token: 단어
        if token.pos_ == "ADJ" and token.head.pos_ in ["NOUN", "PROPN"]: #token.pos_: "이 단어의 품사는 무엇인가" #taken.head: token이 꾸며주는 단어
            pairs.append((token.text, token.head.text))


    return pairs #최종 반환


In [ ]:
# ==== 0) 준비: 라이브러리 ====
import os
import spacy
from tqdm import tqdm
import pandas as pd

# ==== 1) GPU 우선 사용 ====
# CuPy가 설치돼 있고 GPU가 가능하면 True가 떨어짐. 아니면 CPU로 자동 폴백.
if spacy.prefer_gpu():
    print("✅ Using GPU")
else:
    print("⚠️ GPU not available -> Using CPU")

# ==== 2) 모델 로드: 불필요한 컴포넌트 끄기 (속도↑) ====
nlp = spacy.load("en_core_web_sm", disable=["ner"])  # NER 비활성화
# 필요한 파이프만 더 줄이고 싶으면:
# nlp.select_pipes(enable=["tok2vec","tagger","parser","attribute_ruler","lemmatizer"])

# ==== 3) 헬퍼: 한 문서에서 (형용사 -> 명사/고유명사) 쌍 추출 ====
def extract_pairs_from_doc(doc):
    pairs = []
    for tok in doc:
        # 전형적인 '형용사→명사' 수식관계(amod)만 골라서 정확도+속도 ↑
        if tok.pos_ == "ADJ" and tok.dep_ == "amod" and tok.head.pos_ in ("NOUN", "PROPN"):
            # 가벼운 전처리: 소문자 표준화/표제어(lemma) 사용
            pairs.append((tok.lemma_.lower(), tok.head.lemma_.lower()))
    return pairs

# ==== 4) 입력 텍스트 준비 ====
low_score_df = df[df['Score_Difference'] < 0].copy()
texts_series = low_score_df['Cleaned_Negative_Review'].astype(str).str.strip()
texts = texts_series[texts_series != ""].tolist()  # tqdm ETA 정확도를 위해 리스트로

# ==== 5) 배치/병렬 파이프 + tqdm 진행바 ====
batch_size = 256
workers = max(1, (os.cpu_count() or 4) - 1)




⚠️ GPU not available -> Using CPU


In [ ]:
all_pairs = []
for doc in tqdm(nlp.pipe(texts, batch_size=batch_size, n_process=workers),
                total=len(texts), desc="Processing reviews"):
    all_pairs.extend(extract_pairs_from_doc(doc))

print(f"✅ Done. Extracted {len(all_pairs):,} (adj, noun) pairs.")

# ==== 6) 저장: 원시 쌍 + 빈도 집계 ====
pairs_df = pd.DataFrame(all_pairs, columns=["adj", "noun"])
pairs_df.to_csv("noun_adj_pairs_raw.csv", index=False, encoding="utf-8-sig")

freq_df = (pairs_df
           .value_counts(["adj","noun"])
           .rename("count")
           .reset_index()
           .sort_values("count", ascending=False))
freq_df.to_csv("noun_adj_pairs_freq.csv", index=False, encoding="utf-8-sig")

print("💾 Saved:")
print(" - noun_adj_pairs_raw.csv   (모든 쌍)")
print(" - noun_adj_pairs_freq.csv  (빈도 집계 상위부터)")

Processing reviews: 100%|██████████| 183697/183697 [10:07<00:00, 302.37it/s]


✅ Done. Extracted 288,878 (adj, noun) pairs.
💾 Saved:
 - noun_adj_pairs_raw.csv   (모든 쌍)
 - noun_adj_pairs_freq.csv  (빈도 집계 상위부터)


2차 추출

In [ ]:
import pandas as pd

# 1. noun_adj_pairs_freq.csv 파일 불러오기
pairs_df = pd.read_csv('noun_adj_pairs_freq.csv')

# 2. 긍정적인 의미의 형용사 리스트
positive_adjectives = [
    'great', 'nice', 'friendly', 'excellent', 'fantastic',
    'convenient', 'quick', 'perfect', 'spacious', 'helpful']

# 3. 긍정 형용사가 포함된 쌍을 제외
filtered_pairs_df = pairs_df[~pairs_df['adj'].isin(positive_adjectives)]
most_common_filtered_pairs = filtered_pairs_df.sort_values(by='count', ascending=False).head(20)

# 필터링 후 데이터 개수 확인
print(f"필터링 후 남은 쌍 개수: {len(filtered_pairs_df):,}")
print(most_common_filtered_pairs)

필터링 후 남은 쌍 개수: 70,011
         adj   noun  count
0      small   room   7939
1     double   room   2056
2     double    bed   2009
3        hot  water   1430
4     single    bed   1274
5       next    day   1265
6       mini    bar   1241
7       tiny   room   1199
8      first  night   1164
9      other  hotel   1125
10     other   room   1104
11     front   desk   1103
12      next   door   1089
13    single   room    997
14      twin   room    988
15  superior   room    984
16  standard   room    959
17    little    bit    953
18     small    bed    901
19     first   room    833


중립적인 단어도

In [ ]:
# 추가로 제거할 중립적/긍정적 단어 리스트
additional_neutral_words = [
    'hot', 'free', 'next', 'first', 'double', 'little', 'only', 'other',
    'single', 'superior', 'twin', 'old', 'real', 'much', 'enough', 'several'
]

#필터링 된 변수에서 추가 데이터들을 제거
final_filtered_df = filtered_pairs_df[~filtered_pairs_df['adj'].isin(additional_neutral_words)]

# 최종 필터링된 결과에서 가장 많이 등장하는 쌍 확인
most_common_final_pairs = final_filtered_df.sort_values(by='count', ascending=False).head(20)


print(f"최종 필터링 후 남은 쌍 개수: {len(final_filtered_df):,}")
print(most_common_final_pairs)

최종 필터링 후 남은 쌍 개수: 64,402
              adj        noun  count
0           small        room   7939
6            mini         bar   1241
7            tiny        room   1199
11          front        desk   1103
16       standard        room    959
18          small         bed    901
20          small    bathroom    809
22           poor     service    770
23          noisy        room    748
24           long        time    747
25           poor   breakfast    691
26      executive        room    680
28      expensive   breakfast    663
30            bad  experience    618
34            big        room    525
37  uncomfortable         bed    508
39           poor     quality    499
38         second         day    499
42            bad       smell    473
43          early     morning    466


------------------------------------------------

# 추론통계 (카이제곱)

In [ ]:
import pandas as pd
from collections import Counter
import re
import string
import spacy
import os
from tqdm import tqdm

# 1. 데이터 로드 및 전처리
df = pd.read_csv('/content/drive/MyDrive/Hotel_Reviews.csv')
def clean_text(text):
    if not isinstance(text, str):
        return ''
    if 'no positive' in text.lower() or 'no negative' in text.lower():
        return ''
    text = re.sub(r'\[.*?\]', '', text)
    text = re.sub(r'https?://\S+|www\.\S+', '', text)
    text = re.sub(r'[%s]' % re.escape(string.punctuation), '', text)
    text = re.sub(r'\w*\d\w*', '', text)
    text = re.sub(r'\n', '', text)
    text = re.sub(r'\s+', ' ', text).strip()
    text = text.lower()
    return text

# Positive_Review 컬럼을 전처리하여 Cleaned_Positive_Review 컬럼 생성
df['Cleaned_Positive_Review'] = df['Positive_Review'].apply(clean_text)

# Negative_Review 컬럼도 전처리
df['Cleaned_Negative_Review'] = df['Negative_Review'].apply(clean_text)

In [ ]:
from scipy import stats
from collections import Counter
import spacy

pairs_df = pd.read_csv('noun_adj_pairs_freq.csv')
positive_adjectives = [
    'great', 'nice', 'friendly', 'excellent', 'fantastic',
    'convenient', 'quick', 'perfect', 'spacious', 'helpful',
    'good', 'clean', 'comfortable', 'hot', 'free', 'next',
    'first', 'double', 'little', 'only', 'other', 'single',
    'superior', 'twin', 'old', 'real', 'much', 'enough', 'several'
]

# 2. 분석 대상 상위 20개 쌍 리스트 생성
top_20_pairs = [tuple(row) for row in most_common_filtered_pairs[['adj', 'noun']].values]


# 3. '핵심 쌍' 포함 여부 변수 생성 (전체 데이터에 적용)
def has_target_pair_efficient(text):
    if not isinstance(text, str):
        return False
    text_words = set(text.split())
    for adj, noun in top_20_pairs:
        if adj in text_words and noun in text_words:
            return True
    return False

# 긍정 리뷰 컬럼에 함수 적용
df['Has_Target_Pair'] = df['Cleaned_Positive_Review'].apply(has_target_pair_efficient)

# 4. 교차표 생성 및 카이제곱 검정 실행
# 'Reviewer_Score'와 'Average_Score' 컬럼을 바탕으로 평점 범주를 생성합니다.
df['Score_Category'] = df['Reviewer_Score'].apply(lambda x: 'High' if x >= df['Average_Score'].mean() else 'Low')

contingency_table = pd.crosstab(df['Has_Target_Pair'], df['Score_Category'])
chi2, p_value, dof, expected = stats.chi2_contingency(contingency_table)

print("\n---")
print("### 카이제곱 검정 결과")
print(f"Chi-square Statistic (카이제곱 통계량): {chi2:.2f}")
print(f"p-value (p-값): {p_value:.10f}")


---
### 카이제곱 검정 결과
Chi-square Statistic (카이제곱 통계량): 721.42
p-value (p-값): 0.0000000000


In [ ]:
# 4. 교차표 생성 및 카이제곱 검정 실행.
df['Score_Category'] = df['Reviewer_Score'].apply(lambda x: 'High' if x >= df['Average_Score'].mean() else 'Low')

contingency_table = pd.crosstab(df['Has_Target_Pair'], df['Score_Category'])
chi2, p_value, dof, expected = stats.chi2_contingency(contingency_table)

print("### 핵심 쌍 언급 여부와 평점 범주의 교차표 (관측값)")
print(contingency_table)

print("\n---")
print("### 기대 빈도표 (두 변수가 독립일 경우)")
expected_df = pd.DataFrame(expected, index=contingency_table.index, columns=contingency_table.columns)
print(expected_df.round(2))

### 핵심 쌍 언급 여부와 평점 범주의 교차표 (관측값)
Score_Category     High     Low
Has_Target_Pair                
False            277924  213227
True              16050    8537

---
### 기대 빈도표 (두 변수가 독립일 경우)
Score_Category        High        Low
Has_Target_Pair                      
False            279959.25  211191.75
True              14014.75   10572.25


```
[임의적으로]
귀무가설 (H0): 핵심 불만족 쌍의 존재와 낮은 평점 사이에는 통계적으로 유의미한 관계가 없다. (우연히 발생한 관계일 뿐이다)

대립가설 (H1): 핵심 불만족 쌍의 존재와 낮은 평점 사이에는 통계적으로 유의미한 관계가 있다. (핵심 쌍이 낮은 평점의 원인이다)


p-value가 0이므로, 핵심 불만족 쌍이 언급되는 것과 낮은 평점을 주는 것 사이에는 유의미한 관계가 있다.

부정적인 리뷰쌍 x +  높은 점수 : 278,666
부정적인 리뷰쌍 ㅇ + 높은 점수 : 15,308
부정적인 리뷰쌍 x +  낮은 점수 :  212,940
부정적인 리뷰쌍 ㅇ + 낮은 점수 :  8,824



따라서 부정적인 불만족 단어쌍은  무조건 평점을 낮추는 것이 아니라, '높은 점수를 주면서도 불만 사항을 언급하는' 복합적인 형태로 나타난다.

또한, 단순한 불만 사항도 고객이 만족하면 상쇄될 수 있다!!!


```

---------------------------------------------

In [ ]:
import pandas as pd
from collections import Counter

# 1. noun_adj_pairs_freq.csv 파일 불러오기
freq_df = pd.read_csv('noun_adj_pairs_freq.csv')

# 2. 빈도수 40 기준으로 쌍 분류
pairs_over_40 = freq_df[freq_df['count'] > 40]
pairs_below_40 = freq_df[freq_df['count'] <= 40]

print("### 빈도수 40 초과 쌍 개수:", len(pairs_over_40))
print("### 빈도수 40 이하 쌍 개수:", len(pairs_below_40))

# 상위 20개 쌍과 개수를 보고 싶다면
print("\n### 빈도수 40 초과 쌍 (상위 20개):")
print(pairs_over_40.head(20).to_string())

### 빈도수 40 초과 쌍 개수: 843
### 빈도수 40 이하 쌍 개수: 70868

### 빈도수 40 초과 쌍 (상위 20개):
         adj   noun  count
0      small   room   7939
1     double   room   2056
2     double    bed   2009
3        hot  water   1430
4     single    bed   1274
5       next    day   1265
6       mini    bar   1241
7       tiny   room   1199
8      first  night   1164
9      other  hotel   1125
10     other   room   1104
11     front   desk   1103
12      next   door   1089
13    single   room    997
14      twin   room    988
15  superior   room    984
16  standard   room    959
17    little    bit    953
18     small    bed    901
19     first   room    833


데이터 갯수가 702868개 정도만 나오는 이유!

- 1차 필터에서 감정적이지 않은 형용사 - 명사 쌍이 추출
- 2차 필터에서 good, nice, first 등 중립 단어쌍 제거

- 형용사가 명사를 수식하는 관계만 가진 단어쌍들만 남김
- hotel, stay, would같이 감정과 관련 없는 단어들을 제거하고, 오직 감정과 불만을 표현하는 쌍에만 집중

[인사이트]
small room, hot water 와 같이 중립적이거나 부정적으로 해석될 수 있는 단어들은 감정적인 단어 뿐만이 아니라 객관적인 사실을 묘사하는 단어들이 매우 자주 사용되었음.

긍정 단어를 찾아보자!!!

In [ ]:
#코드 반복
import pandas as pd
from collections import Counter
import re
import string
import spacy
import os
from tqdm import tqdm

# 1. 데이터 로드 및 전처리
df = pd.read_csv('/content/drive/MyDrive/Hotel_Reviews.csv')
def clean_text(text):
    if not isinstance(text, str):
        return ''
    if 'no positive' in text.lower() or 'no negative' in text.lower():
        return ''
    text = re.sub(r'\[.*?\]', '', text)
    text = re.sub(r'https?://\S+|www\.\S+', '', text)
    text = re.sub(r'[%s]' % re.escape(string.punctuation), '', text)
    text = re.sub(r'\w*\d\w*', '', text)
    text = re.sub(r'\n', '', text)
    text = re.sub(r'\s+', ' ', text).strip()
    text = text.lower()
    return text

df['Cleaned_Positive_Review'] = df['Positive_Review'].apply(clean_text)

# 2. 명사-형용사 쌍 추출
nlp = spacy.load("en_core_web_sm")
def extract_pairs_from_doc(doc):
    pairs = []
    for tok in doc:
        if tok.pos_ == "ADJ" and tok.dep_ == "amod" and tok.head.pos_ in ("NOUN", "PROPN"):
            pairs.append((tok.lemma_.lower(), tok.head.lemma_.lower()))
    return pairs

# '높은 평점' 리뷰만 추출합니다.
high_score_df = df[df['Reviewer_Score'] >= df['Average_Score']].copy()
texts_series = high_score_df['Cleaned_Positive_Review'].astype(str).str.strip()
texts = texts_series[texts_series != ""].tolist()
batch_size = 256
workers = max(1, (os.cpu_count() or 4) - 1)

all_positive_pairs = []
for doc in tqdm(nlp.pipe(texts, batch_size=batch_size, n_process=workers),
                 total=len(texts), desc="Processing positive reviews"):
    all_positive_pairs.extend(extract_pairs_from_doc(doc))

# 3. 빈도수 계산 및 저장
positive_pairs_freq = Counter(all_positive_pairs).most_common()

positive_pairs_df = pd.DataFrame(positive_pairs_freq, columns=['adj_noun_pair', 'count'])
positive_pairs_df[['adj', 'noun']] = pd.DataFrame(positive_pairs_df['adj_noun_pair'].tolist())
positive_pairs_df = positive_pairs_df.drop(columns=['adj_noun_pair'])
positive_pairs_df = positive_pairs_df[['adj', 'noun', 'count']]
positive_pairs_df.to_csv("positive_noun_adj_pairs_freq.csv", index=False, encoding="utf-8-sig")

Processing positive reviews: 100%|██████████| 304330/304330 [16:19<00:00, 310.66it/s]


In [ ]:
import pandas as pd
from collections import Counter

# 1. noun_adj_pairs_freq.csv 파일 불러오기
freq_df = pd.read_csv('/content/positive_noun_adj_pairs_freq.csv')

# 2. 빈도수 40 기준으로 쌍 분류
pairs_over_40 = freq_df[freq_df['count'] > 40]
pairs_below_40 = freq_df[freq_df['count'] <= 40]

print("### 빈도수 40 초과 쌍 개수:", len(pairs_over_40))
print("### 빈도수 40 이하 쌍 개수:", len(pairs_below_40))

# 상위 20개 쌍과 개수를 보고 싶다면
print("\n### 빈도수 40 초과 쌍 (상위 20개):")
print(pairs_over_40.head(20).to_string())

[인사이트]
높은 점수를 준 고객들은 직접적으로 만족감을 표현하는 단어를 주로 사용하였음.

---------------------------------------------

In [ ]:
#9/18일 할 일


# 유니크 밸류
#고유한 명사 고유한 형용사 각각 (네거/ 긍정) 부정에서 긍정이 제거 된거 제거 되지 않은거 ,,, 긍정에서 부정이 제거된거 제거되지 않은거

1. 부정 리뷰의 고유 명사 수

2. 부정 리뷰의 고유 형용사 수

3. 긍정 리뷰의 고유 명사 수

4. 긍정 리뷰의 고유 형용사 수

5. 부정 리뷰의 모든 쌍

6. 부정 리뷰에서 긍정 형용사 쌍을 제거한 수

7. 긍정 리뷰의 모든 쌍

8. 긍정 리뷰에서 부정 형용사 쌍을 제거한 수

In [ ]:
import pandas as pd
from collections import Counter
import re
import string
import spacy
import os
from tqdm import tqdm

df['Cleaned_Positive_Review'] = df['Positive_Review'].apply(clean_text)
df['Cleaned_Negative_Review'] = df['Negative_Review'].apply(clean_text)

# 2. 명사, 형용사, 쌍 추출 함수 정의
nlp = spacy.load("en_core_web_sm")

def extract_adjectives_from_doc(doc):
    return [tok.lemma_.lower() for tok in doc if tok.pos_ == "ADJ"]

def extract_nouns_from_doc(doc):
    return [tok.lemma_.lower() for tok in doc if tok.pos_ in ("NOUN", "PROPN")]

def extract_pairs_from_doc(doc):
    pairs = []
    for tok in doc:
        if tok.pos_ == "ADJ" and tok.dep_ == "amod" and tok.head.pos_ in ("NOUN", "PROPN"):
            pairs.append((tok.lemma_.lower(), tok.head.lemma_.lower()))
    return pairs

# 3. 긍정/부정 형용사 리스트 정의
positive_adjectives = [
    'great', 'nice', 'friendly', 'excellent', 'fantastic', 'convenient',
    'quick', 'perfect', 'spacious', 'helpful', 'good', 'clean', 'comfortable'
]
negative_adjectives = [
    'small', 'uncomfortable', 'poor', 'noisy', 'bad', 'dirty',
    'tiny', 'rude', 'broken', 'slow', 'old'
]

In [ ]:
# 4. 리뷰별 데이터 추출 및 집계
positive_texts = df['Cleaned_Positive_Review'].astype(str).str.strip().tolist()
negative_texts = df['Cleaned_Negative_Review'].astype(str).str.strip().tolist()
workers = max(1, (os.cpu_count() or 4) - 1)

all_positive_pairs = []
all_positive_adjectives = []
all_positive_nouns = []
for doc in tqdm(nlp.pipe(positive_texts, n_process=workers), total=len(positive_texts), desc="Processing positive reviews"):
    all_positive_pairs.extend(extract_pairs_from_doc(doc))
    all_positive_adjectives.extend(extract_adjectives_from_doc(doc))
    all_positive_nouns.extend(extract_nouns_from_doc(doc))

all_negative_pairs = []
all_negative_adjectives = []
all_negative_nouns = []
for doc in tqdm(nlp.pipe(negative_texts, n_process=workers), total=len(negative_texts), desc="Processing negative reviews"):
    all_negative_pairs.extend(extract_pairs_from_doc(doc))
    all_negative_adjectives.extend(extract_adjectives_from_doc(doc))
    all_negative_nouns.extend(extract_nouns_from_doc(doc))

In [ ]:
# 1) 유니크한 명사/형용사 개수
unique_positive_nouns = len(set(all_positive_nouns))
unique_positive_adjectives = len(set(all_positive_adjectives))
unique_negative_nouns = len(set(all_negative_nouns))
unique_negative_adjectives = len(set(all_negative_adjectives))

# 2) 쌍 데이터프레임 생성
positive_pairs_df = pd.DataFrame(all_positive_pairs, columns=['adj', 'noun'])
negative_pairs_df = pd.DataFrame(all_negative_pairs, columns=['adj', 'noun'])

# 3) 필터링 조건에 따른 쌍 개수
# 부정에서 긍정 형용사 제거된 쌍
negative_filtered_by_positive = negative_pairs_df[~negative_pairs_df['adj'].isin(positive_adjectives)]
# 긍정에서 부정 형용사 제거된 쌍
positive_filtered_by_negative = positive_pairs_df[~positive_pairs_df['adj'].isin(negative_adjectives)]

In [ ]:
### 최종 결과 출력 ###
print("\n### 1. 고유한 명사 및 형용사 수 (긍정/부정)")
print(f"1) 긍정 리뷰의 고유한 명사 수: {unique_positive_nouns:,}")
print(f"2) 긍정 리뷰의 고유한 형용사 수: {unique_positive_adjectives:,}")
print(f"3) 부정 리뷰의 고유한 명사 수: {unique_negative_nouns:,}")
print(f"4) 부정 리뷰의 고유한 형용사 수: {unique_negative_adjectives:,}")

print("\n### 2. 필터링 전/후 쌍 개수")
print(f"5) 부정 리뷰의 모든 쌍 (제거되지 않음): {len(negative_pairs_df):,}")
print(f"6) 부정 리뷰에서 긍정 형용사 제거 후 쌍: {len(negative_filtered_by_positive):,}")
print(f"7) 긍정 리뷰의 모든 쌍 (제거되지 않음): {len(positive_pairs_df):,}")
print(f"8) 긍정 리뷰에서 부정 형용사 제거 후 쌍: {len(positive_filtered_by_negative):,}")

7,8번 개수가 80만개 정도 되는이유

- 리뷰 텍스트 하나하나를 처리해서 그 안의 모든 형용사- 명사 쌍을 추출하였는데, 한 리뷰에서 여러 개의 쌍이 추출되었기 때문!!

- 긍정 리뷰의 경우 고객들은 만족한 여러 이유를 나열하는 하고, 상세히 묘사하는 경향이 있음

[ 예:  The room was beautiful and very clean and quiet The staff were very friendly Has free wine and fruit on arrival Fresh flowers every day and a free newspaper delivered We loved the compimentary slippers and dressing gown Everything was so nice from the t v to the shower it was a double and very powerful Plus a room safe and brilliant selection of coffee tea and biscuits Free shampoo shower gel body lotion and shower cap Even had an emergency pack of cotton buds and cotton wool pads and a sewing kit Also a hairdryer that worked well Loved every minute ] --> 따라서 다양한 쌍이 추출된다.


- 부정 리뷰의 경우 긍정리뷰에 비해 명확하고 짧게 불만 사항을 언급하는 경향이 많다.

[예1: The bathroom was very small and had no actual shower cubicle or tray just a shower curtain and a drain It worked well enough but the addition of a shower tray would have made it much more pleasant to use ]

[예2: The bedrooms are very tired and a bit grubby looking Air con and extractor were broken ]

--------------------------------

 9/19

 - Z-점수를 통해 '평균에서 가장 크게 벗어난' 극단적인 리뷰(이상치)를 찾아보자!

-  Score_difference 컬럼 중앙값 절대 편차를 이용하여 z점수 계산하기

In [ ]:
import pandas as pd
import numpy as np
from scipy import stats

df = pd.read_csv('Hotel_Reviews.csv')

df['Score_Difference'] = df['Reviewer_Score'] - df['Average_Score']

# 중앙값 계산
median_score_diff = df['Score_Difference'].median()

# 절대 편차(MAD) 계산
mad = np.median(np.abs(df['Score_Difference'] - median_score_diff))

# 수정된 Z-점수(Modified Z-score) 계산
# MAD가 0인 경우를 방지하기 .. 0으로 나누는 오류를 막기
if mad == 0:
    df['Modified_Z_Score'] = 0
else:
    df['Modified_Z_Score'] = 0.6745 * (df['Score_Difference'] - median_score_diff) / mad

print("Score_Difference 통계량")
print(df['Score_Difference'].describe())
print("\n수정된 Z점수 통계량")
print(df['Modified_Z_Score'].describe())

[카이제곱 통계내기]


z점수에서 알게 된 이상치들이 리뷰 내 단어쌍과 통계적으로 유의미한가?!

In [ ]:
#이상치 기준 z점수가 +_2.5를 넘는것을 기준!
outlier_threshold = 2.5

In [ ]:
#이상치 여부 판단
df['Is_Outlier'] = df['Modified_Z_Score'].apply(lambda x: abs(x) >= outlier_threshold)

In [ ]:
#핵심쌍 여부 판단을 위한 컬럼 생성
df['Cleaned_Positive_Review'] = df['Positive_Review'].apply(clean_text)
df['Has_Target_Pair'] = df['Cleaned_Positive_Review'].apply(has_target_pair_efficient)

In [ ]:
from scipy import stats

contingency_table = pd.crosstab(df['Is_Outlier'], df['Has_Target_Pair'])
chi2, p_value, dof, expected = stats.chi2_contingency(contingency_table)

print("z-socre 카이제곱 검정 결과")
print(f"Chi-square Statistic (카이제곱 통계량): {chi2:.2f}")
print(f"p-value (p-값): {p_value:.10f}")
print(f"Degrees of Freedom (자유도): {dof}")

```
귀무가설 (H0): 수정된 Z-점수가 ±2.5를 넘는 이상치 리뷰와 핵심 불만족 쌍의 언급 여부는 서로 독립적이다. (관계가 없다)

대립가설 (H1): 수정된 Z-점수가 ±2.5를 넘는 이상치 리뷰와 핵심 불만족 쌍의 언급 여부는 서로 관련이 있다.

p-value (p-값): 0.0000000000
수정된 z점수를 기준으로 한 평점 이상치와 특정 단어쌍의 언급 여부 사이에는 통계적으로 유의미한 관계가 있음.

평점의 이상치(단순히 점수가 낮거나 높은 리뷰가 아닌, 평균에서 크게 벗어난 극단적인 평점을 가진 리뷰)에는 특정 단어 쌍이 포함되어 있을 가능성이 매우 높다!

[인사이트] 평점의 이상치를 유발하는 핵심적인 문제를 파악하고 우리가 호텔을 지을 때는 이를 염두해두어야겠다.

